Implementing HTTP consumer API

shubhamvashisht opened the pull request.

shubhamvashisht committed:
added util class for extracting path parameters

shubhamvashisht committed:
added consumer creation

shubhamvashisht committed:
added subscription

ppatierno reviewed (commented):
@shubhamvashisht I'm reviewing the PR but this is the first thing that came to my mind. You are using the AMQP approach for HTTP as well. AMQP is a stateful protocol with multiplexing : a client creates a connection and then uses links (receiver/sender) for receiving/sending messages from/to Kafka. For each connection there are : 1 source (a producer is shared across sender links) and N sinks (one for each receiver link). When the AMQP client closes the connection, the source and sinks (always running) are deleted. With HTTP the approach is different. An HTTP client connects asking for create a Kafka consumer, so a sink, and after that it disconnect. In your code, on disconnection, the sink is deleted so we lose the Kafka consumer. It means that when an HTTP client connect for subscribing, the sink isn't there anymore. What I mean is that while in the AMQP case the sinks are related to a connection (because to links inside the living connection), with HTTP the sinks are not related to connections. Maybe the Map<conn, endpoint> cannot  work for the HTTP case and we need something different.

ppatierno commented on src/main/java/io/strimzi/kafka/bridge/http/HttpBridge.java:
```diff
@@ -147,6 +149,36 @@ private void processRequests(HttpServerRequest httpServerRequest) {
+                //add sink to list
+                this.endpoints.get(httpServerRequest.connection()).getSinks().add(sink);
+
+                sink.createConsumer(new HttpEndpoint(httpServerRequest));
```
instead of having this "ad hoc" call on `sink` I would always use the pattern to call `sink.handle` passing the `HttpEndpoint`. The `handle` method inside the `sink` (which is empty in your PR) is then in charge to understand (one more time) that it's `CREATE` or `SUBSCRIBE` or `POLL` (or whatever when you'll implement getting messages) and extracting parameters and then calling `initConsumer`, `subscribe` that are in the `SinkBridgeEndpoint` base class (that you should change a little bit to avoid registering the handler for getting messages because we need poll).

ppatierno commented on src/main/java/io/strimzi/kafka/bridge/http/HttpBridge.java:
```diff
@@ -147,6 +149,36 @@ private void processRequests(HttpServerRequest httpServerRequest) {
+                        .findFirst()
+                        .orElseThrow(() -> new RuntimeException("No consumer found with given instance-id"));
+
+                subscribeSink.subscribeToTopic(new HttpEndpoint(httpServerRequest));
```
Same as above, calling the `handle` method on the `subscribeSink` (call it just `sink`) and leaving it to do the work.

ppatierno commented on src/main/java/io/strimzi/kafka/bridge/BridgeConsumerEndpoint.java:
```diff
@@ -0,0 +1,30 @@
+/**
+ * Interface for Sink Endpoints contains consumer specific methods
+ */
+public interface BridgeConsumerEndpoint {
```
Not sure we need this class and if `initConsumer`, `subscribe` can be reused in some way. Maybe we could need a new one for polling but maybe not on this interface.

ppatierno commented:
I added few comments about using the `sink.handle` for handling requests which is something that you are already doing on the source side `HttpSourceBridgeEndpoint.handle`

ppatierno commented on src/main/java/io/strimzi/kafka/bridge/BridgeConsumerEndpoint.java:
```diff
@@ -0,0 +1,30 @@
+
+    void subscribeToTopic(Endpoint<?> endpoint);
+
+    void consume(Endpoint<?> endpoint);
```
In any case we don't need a poll/consume method in an interface because it's not going to be called from outside. If we use the `sink.handle` for handling HTTP client requests (as suggested in the following comments), it will just be a one more protected method in the base sink class for calling a poll on Vert.x consumer.

shubhamvashisht committed:
Changed "adhoc" calls to sink.() to single sink.handle().

Reformated HttpSinkEnpoint and creation, subscription implementation.

added Separate Map<> for sinkEndpoints. i.e. sinkEndpoints lives after connection closes.

shubhamvashisht committed:
added consuming records

ppatierno commented on src/main/java/io/strimzi/kafka/bridge/SinkBridgeEndpoint.java:
```diff
@@ -74,9 +74,11 @@
 
     //unique id assigned to every consumer during its creation.
-    public String consumerInstanceId;
+    protected String consumerInstanceId;
```
the consumer-id is something needed just for the HTTP bridge so I think that we can move this field to the `HttpSinkBridgeEndpoint` class.

ppatierno commented on src/main/java/io/strimzi/kafka/bridge/SinkBridgeEndpoint.java:
```diff
@@ -74,9 +74,11 @@
+    protected String consumerInstanceId;
 
-    public String consumerBaseUri;
+    protected String consumerBaseUri;
```
the same as for the consumer-id, we need this just for the HTTP bridge.

ppatierno commented on src/main/java/io/strimzi/kafka/bridge/SinkBridgeEndpoint.java:
```diff
@@ -143,14 +145,19 @@ protected void initConsumer() {
         this.consumer = KafkaConsumer.create(this.vertx, props);
+
+        if (shouldAttachBatchHandler)
         this.consumer.batchHandler(this::handleKafkaBatch);
```
wrong indentation

ppatierno commented on src/main/java/io/strimzi/kafka/bridge/SinkBridgeEndpoint.java:
```diff
@@ -266,6 +273,7 @@ private void automaticPartitionAssignment() {
             }
 
+            if (shouldAttachSubscriberHandler)
             this.consumer.handler(this::handleKafkaRecord);
```
wrong indentation

ppatierno commented on src/main/java/io/strimzi/kafka/bridge/SinkBridgeEndpoint.java:
```diff
@@ -276,6 +284,8 @@ private void automaticPartitionAssignment() {
         this.handlePartitionsAssigned(partitions);
+
+        if (shouldAttachSubscriberHandler)
         this.consumer.handler(this::handleKafkaRecord);
```
wrong indentation

ppatierno commented on src/main/java/io/strimzi/kafka/bridge/BridgeEndpoint.java:
```diff
@@ -39,6 +39,13 @@
+     * @param endpoint	Remote protocol endpoint to handle
+     * @param handler handler for result
+     */
+	void handle(Endpoint<?> endpoint, Handler<String> handler);
```
the `String` is because we know that for the HTTP case we have a consumer-id but for example it doesn't make sense for the AMQP bridge. I know that we are not using this `handle` for the AMQP bridge but in any case we are forcing to an handler with `String` as return type. Can we be more generic like for the other `endpoint` parameter ?

ppatierno commented on src/main/java/io/strimzi/kafka/bridge/http/HttpBridge.java:
```diff
@@ -46,6 +45,8 @@
 
     private Map<HttpConnection, ConnectionEndpoint> endpoints;
```
Do we still need this collection? What's the purpose now?

ppatierno commented on src/main/java/io/strimzi/kafka/bridge/http/HttpBridge.java:
```diff
@@ -147,6 +156,38 @@ private void processRequests(HttpServerRequest httpServerRequest) {
+                final SinkBridgeEndpoint consumeSink = this.httpSinkEndpoints.get(consumeInstanceId);
+
+                consumeSink.handle(new HttpEndpoint(httpServerRequest));
+
```
there is a `break` missing. Anyway, the `CONSUME` and `SUBSCRIBE` cases are quite the same can we collapse them in only one case? i.e. 

```
case SUBSCRIBE:
case CONSUME:
      String instanceId = ....
      ...
      ...
      break;
```

ppatierno commented on src/main/java/io/strimzi/kafka/bridge/http/HttpSinkBridgeEndpoint.java:
```diff
@@ -0,0 +1,174 @@
+                    this.pollTimeOut = Long.parseLong(httpServerRequest.getHeader("timeout"));
+                }
+
+                this.consume();
```
I think we don't need the `setManualRecordBatchHandler` method but we can pass the handler to the `consume` method directly. Or maybe I'm missing something?

ppatierno commented on src/main/java/io/strimzi/kafka/bridge/converter/MessageConverter.java:
```diff
@@ -42,4 +43,11 @@
+	 * @param records KafkaRecords to convert
+	 * @return message
+	 */
+	M toMessages(KafkaConsumerRecords<K, V> records);
```
I don't think that it works. For AMQP for example you cannot have `toMessages` plural and then returning just a single `Message`; it should be at least a collection (i.e. `Collection<M>`). But then it doesn't work well with HTTP where you want a `Buffer` containing all messages and not a `Collection<Buffer>`.
Not sure if it works but you could have `MessageConverter<K, V, M, L>` and having `toMessages` returning `C` where for the AMQP implementatin is a `Collection<M>` and for HTTP `M` and `C` are the same and they are `Buffer`. Wdyt?

shubhamvashisht committed:
Fixed endpoint map for producer.
Fixed identation errors.
Made handle(endpoint,hanler) generic.
Fixed switch cases for identifying request types.
Removed setManualRecordsBatchHandler.
Fixed MessageConverter.

ppatierno commented on src/main/java/io/strimzi/kafka/bridge/amqp/AmqpSinkBridgeEndpoint.java:
```diff
@@ -50,7 +53,7 @@
 	
 	// converter from ConsumerRecord to AMQP message
-	private MessageConverter<K, V, Message> converter;
+	private MessageConverter<K, V, Message, Collection<Buffer>> converter;
```
It needs to be a `Collection<Message>` (you are using `Buffer` for HTTP not AMQP)

ppatierno commented on src/main/java/io/strimzi/kafka/bridge/amqp/converter/AmqpDefaultMessageConverter.java:
```diff
@@ -38,7 +41,7 @@
  * between Kafka record and AMQP message
  */
-public class AmqpDefaultMessageConverter implements MessageConverter<String, byte[], Message> {
+public class AmqpDefaultMessageConverter implements MessageConverter<String, byte[], Message, Collection<Buffer>> {
```
`Collection<Message>`

ppatierno commented on src/main/java/io/strimzi/kafka/bridge/amqp/converter/AmqpDefaultMessageConverter.java:
```diff
@@ -130,4 +133,9 @@ public Message toMessage(String address, KafkaConsumerRecord<String, byte[]> rec
 	}
 
+	@Override
+	public Collection<Buffer> toMessages(KafkaConsumerRecords<String, byte[]> records) {
```
`Collection<Message>`

ppatierno commented on src/main/java/io/strimzi/kafka/bridge/amqp/converter/AmqpJsonMessageConverter.java:
```diff
@@ -18,32 +18,26 @@
  * between Kafka record and AMQP message
  */
-public class AmqpJsonMessageConverter implements MessageConverter<String, byte[], Message> {
+public class AmqpJsonMessageConverter implements MessageConverter<String, byte[], Message, Collection<Buffer>> {
```
`Collection<Message>`

ppatierno commented on src/main/java/io/strimzi/kafka/bridge/amqp/converter/AmqpJsonMessageConverter.java:
```diff
@@ -291,5 +285,10 @@ public Message toMessage(String address, KafkaConsumerRecord<String, byte[]> rec
 
-	
+	@Override
+	public Collection<Buffer> toMessages(KafkaConsumerRecords<String, byte[]> records) {
```
`Collection<Message>`

ppatierno commented on src/main/java/io/strimzi/kafka/bridge/amqp/converter/AmqpRawMessageConverter.java:
```diff
@@ -34,7 +37,7 @@
  * It passes the AMQP message as is (raw bytes) as Kafka record value and vice versa.
  */
-public class AmqpRawMessageConverter implements MessageConverter<String, byte[], Message> {
+public class AmqpRawMessageConverter implements MessageConverter<String, byte[], Message, Collection<Buffer>> {
```
`Collection<Message>`

ppatierno commented on src/main/java/io/strimzi/kafka/bridge/amqp/converter/AmqpRawMessageConverter.java:
```diff
@@ -96,4 +99,9 @@ public Message toMessage(String address, KafkaConsumerRecord<String, byte[]> rec
 	}
 
+	@Override
+	public Collection<Buffer> toMessages(KafkaConsumerRecords<String, byte[]> records) {
```
`Collection<Message>`

ppatierno commented on src/main/java/io/strimzi/kafka/bridge/converter/MessageConverter.java:
```diff
@@ -42,4 +43,11 @@
+	/**
+	 * Converts multiple KafkaConsumerRecord from KafkaConsumerRecords to a message
+	 * @param records KafkaRecords to convert
+	 * @return message
```
it's a list of messages, not a "message"

shubhamvashisht committed:
added consumer deletion upon client request

ppatierno commented on src/main/java/io/strimzi/kafka/bridge/http/HttpBridge.java:
```diff
@@ -173,7 +173,21 @@ private void processRequests(HttpServerRequest httpServerRequest) {
+            case DELETE:
+                String deleteInstanceID = PathParamsExtractor.getConsumerDeletionParams(httpServerRequest).get("instance-id");
+
+                final SinkBridgeEndpoint deleteSinkEndpoint = this.httpSinkEndpoints.get(deleteInstanceID);
```
Should check if client asks for a not existing sink bridge handpoint. If it's null, the sequent instruction will raise an NPE.

shubhamvashisht committed:
added commiting offsets

ppatierno commented on src/main/java/io/strimzi/kafka/bridge/SinkBridgeEndpoint.java:
```diff
@@ -556,4 +564,14 @@ protected void consume(Handler<KafkaConsumerRecords<K, V>> manualRecordBatchHand
+        this.commitOffsetsHandler = commitOffsetsHandler;
+
+        this.consumer.commit(offsetsData, result -> {
+            if (result.succeeded() && this.commitOffsetsHandler != null) {
```
You should handle the case when commit fails so result.succeded() is false. Maybe instead of a Handler<Map....> you should use Handler<AsyncResult<Map...>> so that the AsyncResult can be set to be failed.

shubhamvashisht committed:
added test for consumer and fixed partitionAssignHandler

shubhamvashisht committed:
fixed commit offset handler

shubhamvashisht committed:
added doc for consumer API

shubhamvashisht committed:
added consumer test for receiving message from partition and offset

ppatierno commented on src/main/java/io/strimzi/kafka/bridge/http/HttpSinkBridgeEndpoint.java:
```diff
@@ -133,7 +133,12 @@ public void handle(Endpoint<?> endpoint) {
+                            sendConsumerCommitOffsetResponse(httpServerRequest.response(), true);
+                        } else {
+                            sendConsumerCommitOffsetResponse(httpServerRequest.response(), false);
+                        }
```
why not just passing `status.succeeded()` instead of the `if` and then passing `true` or `false` ? :-)

ppatierno commented on src/main/java/io/strimzi/kafka/bridge/http/HttpBridge.java:
```diff
@@ -131,30 +138,71 @@ private void processRequests(HttpServerRequest httpServerRequest) {
+
+                    this.httpSinkEndpoints.remove(deleteInstanceID);
+                } else {
+                    throw new NullPointerException("no conusmer instance found with this id");
```
you should not throw an exception but replying to the HTTP client with a "404 Not Found" status code even because we want to avoid an NPE but you are raising an NPE :-)

ppatierno commented on doc/HTTP Consumer API.md:
```diff
@@ -0,0 +1,120 @@
+# Consumer API
```
Can you rename this file using underscores and not spaces please? The same for the HTTP producer API doc.

ppatierno commented on doc/HTTP Consumer API.md:
```diff
@@ -0,0 +1,120 @@
+#### Consumer API provides allows you to create a consumer in a consumer group and consume messages from topics and partitions. For each type of request only `JSON` format is supported.
+
+* ### Creating a consumer
+    * #### Create a new consumer instance in the consumer group. Optionally specifying name of     consumer instance. If name is not specified a random name will be assgined.
```
"name of      consumer" too many spaces.

ppatierno replied:
typo "assigned"

ppatierno commented on src/test/java/io/strimzi/kafka/bridge/http/HttpBridgeTest.java:
```diff
@@ -464,4 +465,86 @@ public void receiveSimpleMessageFromPartition(TestContext context) {
+
+        String name = "kafkaconsumer123";
+
+        String baseUri = "http://0.0.0.0:8080/consumers/group1/instances/"+name;
```
it's better using `BRIDGE_PORT` and `BRIDGE_HOST` for building this string.

ppatierno commented on src/test/java/io/strimzi/kafka/bridge/http/HttpBridgeTest.java:
```diff
@@ -303,4 +304,164 @@ public void sendPeriodicMessage(TestContext context){
+
+        String name = "kafkaconsumer123";
+
+        String baseUri = "http://0.0.0.0:8080/consumers/group1/instances/"+name;
```
it's better using `BRIDGE_PORT` and `BRIDGE_HOST` for building this string.

ppatierno commented on src/test/java/io/strimzi/kafka/bridge/http/HttpBridgeTest.java:
```diff
@@ -303,4 +304,164 @@ public void sendPeriodicMessage(TestContext context){
+
+        String name = "kafkaconsumer123";
+
+        String baseUri = "http://0.0.0.0:8080/consumers/group1/instances/"+name;
```
it's better using `BRIDGE_PORT` and `BRIDGE_HOST` for building this string.

ppatierno commented on src/test/java/io/strimzi/kafka/bridge/http/HttpBridgeTest.java:
```diff
@@ -303,4 +304,164 @@ public void sendPeriodicMessage(TestContext context){
+        }).putHeader("Content-length", String.valueOf(subJson.toBuffer().length()))
+                .putHeader("timeout", String.valueOf(1000))
+                .end();
+
```
do we want to end by deleting the consumer?

ppatierno commented on src/test/java/io/strimzi/kafka/bridge/http/HttpBridgeTest.java:
```diff
@@ -303,4 +304,164 @@ public void sendPeriodicMessage(TestContext context){
+        }).putHeader("Content-length", String.valueOf(subJson.toBuffer().length()))
+                .putHeader("timeout", String.valueOf(1000))
+                .end();
+
```
do we want to end by deleting the consumer?

ppatierno commented on src/test/java/io/strimzi/kafka/bridge/http/HttpBridgeTest.java:
```diff
@@ -464,4 +465,86 @@ public void receiveSimpleMessageFromPartition(TestContext context) {
+        }).putHeader("Content-length", String.valueOf(subJson.toBuffer().length()))
+                .putHeader("timeout", String.valueOf(1000))
+                .end();
+        consumeAsync.await();
```
do we want to end by deleting the consumer?

shubhamvashisht committed:
Fixed typos and renamed HTTP_consumer_API.md doc

shubhamvashisht committed:
fixed NPE while deleting consumer

shubhamvashisht committed:
fixed method javadoc

shubhamvashisht committed:
replaced if-else with asyncResult.succeeded

shubhamvashisht committed:
Rename HTTP Producer API.md to HTTP_Producer_API.md

shubhamvashisht committed:
changed baseUri String with BRIDGE_HOST and BRIDGE_PORT

shubhamvashisht committed:
added consumer deletion test

ppatierno commented on src/main/java/io/strimzi/kafka/bridge/http/HttpBridge.java:
```diff
@@ -187,12 +187,15 @@ private void processRequests(HttpServerRequest httpServerRequest) {
+                    httpServerRequest.response()
+                            .setStatusCode(404)
+                            .putHeader("Content-length", String.valueOf("".length()))
+                            .write("")
```
when the body is empty isn't enough calling `end()` avoiding an header with "Content-length" 0 and writing an empty string? Just asking, have you already tried?

ppatierno reviewed (commented):
Almost there but I have a comment. In a lot of parts you are using `putHeader("Content-length",String.valueOf("".length())).write("").end();` are you sure that you cannot use `end` directly without writing an empty body and content-length header to 0?

shubhamvashisht committed:
fixed all the responses

ppatierno commented on src/main/java/io/strimzi/kafka/bridge/SinkBridgeEndpoint.java:
```diff
@@ -521,4 +537,41 @@ private void handleCommit(AsyncResult<Void> commitResult) {
+        }
+    }
+
+    private void handleCommitedOffsets(AsyncResult<Map<TopicPartition, io.vertx.kafka.client.consumer.OffsetAndMetadata>> offsetsData) {
```
This parameter has to be an `Handler<AsyncResult<Map....>` as the related `commitOffsetsHandler` field.

ppatierno commented on src/main/java/io/strimzi/kafka/bridge/SinkBridgeEndpoint.java:
```diff
@@ -521,4 +537,41 @@ private void handleCommit(AsyncResult<Void> commitResult) {
+                        this.handleManualRecords(pollResult.result());
+                    }
+                }
+            }
```
what happens if it doesn't succeed? The HTTP client doesn't get any response. So I guess that the `manualRecordBatchHandler` should be an `Handler<AsyncResult<KafkaConsumer......>` so that the bridge is able to send a HTTP 500 error if it fails.

shubhamvashisht committed:
fixed and handled cases when conusme,commit fails

ppatierno reviewed (approved):
LGTM
